### PR TITLE
Time series forecasting with sklearn regressors

### PR DESCRIPTION
**Pull request by int-chaos**
add sklearn regressors ['lgbm', 'xgboost', 'rf', 'extra_tree'] as learners for ts_forecasting task
- support iterative and direct strategy

warnings and errors for duplicate rows and missing values #297

### CODE REVIEW

File: flaml/model.py [hunk 1]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+        })
+        return space
+
+
+    def __init__(self, task=TS_FORECAST, **params):
+        super().__init__(task, **params)
+        self.hcrystaball_model = None
+
+    def transform_X(self, X):
+        X = X.copy()
```
Comment (sonichi) [inline flaml/model.py:1781]:
```suggestion
```

File: flaml/model.py [hunk 2]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+    @classmethod
+    def search_space(cls, data_size, pred_horizon, **params):
+        space = cls.base_class.search_space(data_size, **params)
+        space.update({
+            "optimize_for_horizon": {
+                "domain": tune.choice([True, False]),
+                "low_cost_init_value": False
+            },
+            "lags": {
+                "domain": tune.quniform(lower=1, upper=data_size[0]-(pred_horizon+1), q=1),
```
Comment (sonichi) [inline flaml/model.py:1769]:
```suggestion
                "domain": tune.randint(lower=1, upper=data_size[0]-pred_horizon),
``

File: flaml/model.py [hunk 3]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+        X_train = self._preprocess(X_train)
+        params = self.params.copy()
+        lags = params.pop("lags")
+        optimize_for_horizon = params.pop("optimize_for_horizon")
+        estimator = flaml_estimator(task="regression", **params)
+        self.hcrystaball_model = get_sklearn_wrapper(estimator.estimator_class)
+        self.hcrystaball_model.lags = int(lags)
+        self.hcrystaball_model.fit(X_train, y_train)
+        if optimize_for_horizon == True:
+            model_li = []
```
Comment (sonichi) [inline flaml/model.py:1805]:
```suggestion
            model_list = []
```

File: flaml/model.py [hunk 4]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+            X_fit, y_fit = self.hcrystaball_model._transform_data_to_tsmodel_input_format(X_train, y_train, kwargs["period"])
+            self.hcrystaball_model.model.set_params(**estimator.params)
+            model = self.hcrystaball_model.model.fit(X_fit, y_fit)
+            self._model = model
+
+
+    def predict(self, X_test):
+        if self._model is not None:
+            X_test = self._preprocess(X_test)
+            X_test = self.transform_X(X_test)
```
Comment (sonichi) [inline flaml/model.py:1822]: The order of these two lines are different from 1795-1796

File: flaml/model.py [hunk 5]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+        params = self.params.copy()
+        lags = params.pop("lags")
+        optimize_for_horizon = params.pop("optimize_for_horizon")
+        estimator = flaml_estimator(task="regression", **params)
+        self.hcrystaball_model = get_sklearn_wrapper(estimator.estimator_class)
+        self.hcrystaball_model.lags = int(lags)
+        self.hcrystaball_model.fit(X_train, y_train)
+        if optimize_for_horizon == True:
+            model_li = []
+            for i in range(1, kwargs["period"]+1):
```
Comment (sonichi) [inline flaml/model.py:1806]: lint

File: flaml/model.py [hunk 6]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+            X_test = self.transform_X(X_test)
+            if isinstance(self._model, list):
+                assert (
+                    len(self._model) == len(X_test)
+                ), "Model is optimized for horizon, length of X_test must be equal to `period`."
+                preds = []
+                for i in range(1, len(self._model) + 1):
+                    X_pred, _ = self.hcrystaball_model._transform_data_to_tsmodel_input_format(X_test.iloc[:i, :])
+                    preds.append(self._model[i - 1].predict(X_pred)[-1])
+                forecast = pd.DataFrame(data=np.asarray(preds).reshape(-1, 1), columns=[self.hcrystaball_model.name], index=X_test.index)
```
Comment (sonichi) [inline flaml/model.py:1831]: lint

File: flaml/model.py [hunk 7]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+            else:
+                X_pred, _ = self.hcrystaball_model._transform_data_to_tsmodel_input_format(X_test)
+                forecast = self._model.predict(X_pred)
+            return forecast
+        else:
+            logger.warning(
+                "Estimator is not fit yet. Please run fit() before predict()."
+            )
+            return np.ones(X_test.shape[0])
+
```
Comment (sonichi) [inline flaml/model.py:1841]:
```suggestion


```

File: flaml/model.py [hunk 8]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+        if len(cols) == 1:
+            ds_col = cols[0]
+            X = pd.DataFrame(index=X[ds_col])
+        elif len(cols) > 1:
+            ds_col = cols[0]
+            exog_cols = cols[1:]
+            X = X[exog_cols].set_index(X[ds_col])
+        return X
+
+    def _fit(self, flaml_estimator, X_train, y_train, budget=None, **kwargs):
```
Comment (sonichi) [inline flaml/model.py:1792]: remove `flaml_estimator` and use `self.base_class`

File: flaml/model.py [hunk 9]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+
+
+class ExtraTrees_TS_Regressor(TS_SKLearn_Regressor):
+    base_class = ExtraTreesEstimator
+
+    def fit(self, X_train, y_train, budget=None, **kwargs):
+        current_time = time.time()
+        self._fit(self.base_class, X_train, y_train, budget=budget, **kwargs)
+        train_time = time.time() - current_time
+        return train_time
```
Comment (sonichi) [inline flaml/model.py:1889]: define fit() only once in TK_SKLearn_Regressor

**Review by sonichi — COMMENTED** [review]
Would you update the notebook and documentation?
If you plan to do it in the next PR, add an issue and assign to yourself.

File: flaml/model.py [hunk 10]
```diff
@@ -1754,6 +1754,141 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+            ds_col = cols[0]
+            X = pd.DataFrame(index=X[ds_col])
+        elif len(cols) > 1:
+            ds_col = cols[0]
+            exog_cols = cols[1:]
+            X = X[exog_cols].set_index(X[ds_col])
+        return X
+
+    def _fit(self, flaml_estimator, X_train, y_train, budget=None, **kwargs):
+        from hcrystalball.wrappers import get_sklearn_wrapper
```
Comment (sonichi) [inline flaml/model.py:1793]: add dependency in setup.py. section [test], [ts_forecast], and [forecast]

File: flaml/model.py (hunk 6 above)
Comment (int-chaos) [reply flaml/model.py:1831]: the Lint test says `.\flaml\model.py:1828:47: E226 missing whitespace around arithmetic operator` but I can't see where it occurs

File: flaml/automl.py [hunk 11]
```diff
@@ -858,6 +881,20 @@ def _validate_data(
...
+                                  f"\n{dataframe.loc[ts_col.duplicated(), :]}")
+                    dataframe = dataframe.copy().drop_duplicates()
+                    warnings.warn(f"Removed duplicate rows based on all columns")
+                    assert not dataframe[[dataframe.columns[0]]].duplicated(), (
+                        "Duplicate timestamp values with differnt values for other columns.")
+                ts_series = pd.to_datetime(dataframe[dataframe.columns[0]])
+                inferred_freq = pd.infer_freq(ts_series)
+                if inferred_freq == None:
+                    import warnings
+                    warnings.warn(f"Missing timestamps detected. To avoid error with estimators, set estimator list to ['prophet']. ")
```
[the remark this answers is not in the document]
Comment (int-chaos) [reply flaml/automl.py:897]: Should I make a new function `def _validate_ts_data()`? If yes, do you have a specific place you want me to define it (in regards to the other functions, like before or after a function)?

File: flaml/automl.py [hunk 12]
```diff
@@ -2094,8 +2124,18 @@ def is_to_reverse_metric(metric, task):
                         "extra_tree",
                         "xgb_limitdepth",
                     ]
-                if "regression" != self._state.task:
+                if TS_FORECAST == self._state.task:
+                    estimator_list.remove("catboost")
+                    estimator_list.remove("xgb_limitdepth")
```
[the remark this answers is not in the document]
Comment (int-chaos) [reply flaml/automl.py:2129]: catboost is removed because it has a `name` parameter, making it incompatible with hcrystalball 

will be testing out xgb_limitdepth

File: flaml/automl.py [hunk 13]
```diff
@@ -813,6 +813,38 @@ def _preprocess(self, X):
...
+                dataframe[[dataframe.columns[0]]].duplicated() is None
+            ), "Duplicate timestamp values with differnt values for other columns."
+        ts_series = pd.to_datetime(dataframe[dataframe.columns[0]])
+        inferred_freq = pd.infer_freq(ts_series)
+        if inferred_freq == None:
+            logger.warning(
+                f"Missing timestamps detected. To avoid error with estimators, set estimator list to ['prophet']. "
+            )
+        if y_train_all is not None:
+            return dataframe.iloc[:,:-1], dataframe.iloc[:, -1]
```
Comment (sonichi) [inline flaml/automl.py:845]: lint

File: setup.py [hunk 14]
```diff
@@ -77,8 +78,8 @@
             "vowpalwabbit",
         ],
         "nlp": ["transformers", "datasets", "torch", "nltk", "rouge_score"],
-        "ts_forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2"],
-        "forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2"],
+        "ts_forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball"],
+        "forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball"],
```
Comment (sonichi) [inline setup.py:82]: fix the version number of hcrystalball

**Review by sonichi — COMMENTED** [review]
Check the build result for lint issues.

File: flaml/automl.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/automl.py]: Please add a comment about catboost

File: flaml/model.py (hunk 6 above)
Comment (sonichi) [reply flaml/model.py]: :

File: flaml/automl.py [hunk 15]
```diff
@@ -827,7 +827,7 @@ def _validate_ts_data(
             dataframe = dataframe.join(y_df)
         if any(ts_col.duplicated()):
             logger.warning(
-                f"Duplicate timestamp values found in timestamp column. "
+                "Duplicate timestamp values found in timestamp column. "
                 f"\n{dataframe.loc[ts_col.duplicated(), dataframe.columns[0]]}"
```
Comment (sonichi) [inline flaml/automl.py:831]: don't call `ts_col.duplicated()` twice.
How long does it take to deduplicate? There may be a potential performance issue for large data size.

File: setup.py [hunk 16]
```diff
@@ -60,7 +60,7 @@
             "torch",
             "nltk",
             "rouge_score",
-            "hcrystalball",
+            "hcrystalball>=0.1.10",
```
Comment (sonichi) [inline setup.py:63]:
```suggestion
            "hcrystalball==0.1.10",
```

File: setup.py [hunk 17]
```diff
@@ -78,8 +78,8 @@
             "vowpalwabbit",
         ],
         "nlp": ["transformers", "datasets", "torch", "nltk", "rouge_score"],
-        "ts_forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball"],
-        "forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball"],
+        "ts_forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball>=0.1.10"],
+        "forecast": ["prophet>=1.0.1", "statsmodels>=0.12.2", "hcrystalball>0.1.10"],
```
Comment (sonichi) [inline setup.py:82]: ==

**Review by sonichi — CHANGES_REQUESTED** [review]
Out of date with main.

File: flaml/model.py [hunk 18]
```diff
@@ -1776,6 +1776,122 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
 
 
+class TS_SKLearn_Regressor(SKLearnEstimator):
+    base_class = SKLearnEstimator
+
+    @classmethod
+    def search_space(cls, data_size, pred_horizon, **params):
+        space = cls.base_class.search_space(data_size, **params)
+        space.update({
+            "optimize_for_horizon": {
```
Comment (slhuang) [inline flaml/model.py:1786]: add some comments on what does `optimize_for_horizon` mean?

File: flaml/model.py [hunk 19]
```diff
@@ -1776,6 +1776,122 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+
+        X_train = self.transform_X(X_train)
+        X_train = self._preprocess(X_train)
+        params = self.params.copy()
+        lags = params.pop("lags")
+        optimize_for_horizon = params.pop("optimize_for_horizon")
+        estimator = self.base_class(task="regression", **params)
+        self.hcrystaball_model = get_sklearn_wrapper(estimator.estimator_class)
+        self.hcrystaball_model.lags = int(lags)
+        self.hcrystaball_model.fit(X_train, y_train)
```
Comment (slhuang) [inline flaml/model.py:1824]: what is the difference between line 1824 and line 1830 (1836)?

File: flaml/model.py [hunk 20]
```diff
@@ -1776,6 +1776,122 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
+        if self._model is not None:
+            X_test = self.transform_X(X_test)
+            X_test = self._preprocess(X_test)
+            if isinstance(self._model, list):
+                assert (
+                    len(self._model) == len(X_test)
+                ), "Model is optimized for horizon, length of X_test must be equal to `period`."
+                preds = []
+                for i in range(1, len(self._model) + 1):
+                    X_pred, _ = self.hcrystaball_model._transform_data_to_tsmodel_input_format(X_test.iloc[:i, :])
```
Comment (slhuang) [inline flaml/model.py:1855]: If there exists `lag` in the model, the first row of X_test would not get the corresponding `lag` value?

File: flaml/automl.py [hunk 21]
```diff
@@ -802,6 +809,39 @@ def _preprocess(self, X):
...
+            y_df = pd.DataFrame(y_train_all) if isinstance(y_train_all, pd.Series) else pd.DataFrame(y_train_all, columns=['labels'])
+            dataframe = dataframe.join(y_df)
+        duplicates = ts_col.duplicated()
+        if any(duplicates):
+            logger.warning(
+                "Duplicate timestamp values found in timestamp column. "
+                f"\n{dataframe.loc[duplicates, dataframe.columns[0]]}"
+            )
+            dataframe = dataframe.drop_duplicates()
+            logger.warning(f"Removed duplicate rows based on all columns")
```
[the remark this answers is not in the document]
Comment (int-chaos) [reply flaml/automl.py:831]: `duplicates = ts_col.duplicated()` is used to get the duplicated timestamps (which is the first column of dataframe). If there are any, then the users are warned about them, only the timestamps not the entire row. `dataframe.drop_duplicates()` is used to drop the rows with ALL the same values (including timestamps and label values and exogenous values, if they exist). If nothing is dropped, this means that there was at least one timestamp with different values, which will prompt an error.

File: flaml/automl.py [hunk 22]
```diff
@@ -802,6 +809,39 @@ def _preprocess(self, X):
...
+            assert (
+                dataframe[[dataframe.columns[0]]].duplicated() is None
+            ), "Duplicate timestamp values with differnt values for other columns."
```
Comment (int-chaos) [inline flaml/automl.py:834]:
```suggestion
            ), "Duplicate timestamp values with different values for other columns."
```

File: flaml/automl.py [hunk 23]
```diff
@@ -738,6 +741,10 @@ def predict(
             )
             return None
         X_test = self._preprocess(X_test)
+        # TODO: check to see length of X_test is the same as inputted period
+        # if self._state.task == TS_FORECAST:
+        #     assert self._state.fit_kwargs.get("period") == len(X_test), (
+        #         "length of X_test needs to be the same as `periods`")
```
[the remark this answers is not in the document]
Comment (int-chaos) [reply flaml/automl.py:747]: it is required if `optimize_for_horizon` is true, but I found another way to make sure the `X_test` is the same length as `period`

Comment (int-chaos) [inline flaml/automl.py:747]:
```suggestion
```

File: flaml/model.py (hunk 20 above)
Comment (int-chaos) [reply flaml/model.py:1855]: lag features are handled during `self.hcrystaball_model._transform_data_to_tsmodel_input_format()` I just tested to make sure and yes it would using previous data, I believe.

File: flaml/automl.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (slhuang) [reply flaml/automl.py:827]: I feel you can simply 1. check whether there are duplicates in `dataframe`; if so, `drop_duplicates` and output warning; 2. check whether there still exist duplicates in the timestamp column and output warning.

File: flaml/automl.py [hunk 24]
```diff
@@ -831,7 +827,7 @@ def _validate_ts_data(
             logger.warning(f"Removed duplicate rows based on all columns")
             assert (
```
Comment (slhuang) [inline flaml/automl.py:828]: Please change assert to warning, and add a test to demonstrate that forecasting with regression can handle multiple time-series in the datasets, i.e., with duplicated timestamp but with different values on some other attribute.

File: flaml/model.py (hunk 19 above)
Comment (slhuang) [reply flaml/model.py:1826]: got it. Thanks!

File: flaml/model.py (hunk 20 above)
Comment (slhuang) [reply flaml/model.py:1858]: interesting, so it will use the corresponding previous value from the training data?

File: flaml/automl.py (hunk 24 above)
Comment (int-chaos) [reply flaml/automl.py:828]: We decided that flaml will not support forecasting with regression when there are duplicated timestamps but different values on some other attribute. The only known estimator that was capable of it was prophet and based on a benchmark test, the performance was suboptimal. This is why an error is prompted.

Comment (slhuang) [reply flaml/automl.py:828]: @sonichi

File: flaml/automl.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (int-chaos) [reply flaml/automl.py]: added on local file

File: flaml/automl.py (hunk 24 above)
Comment (sonichi) [reply flaml/automl.py:828]: @int-chaos we won't handle duplicate timestamps when there is only one time series. But we do want to handle multiple time series indicated by categorical columns.

**Review by slhuang — APPROVED** [review]

**Review by sonichi — APPROVED** [review]
